### PR TITLE
Implement RFC4363 dot1qTpFdbTable Q-BRIDGE-MIB

### PR DESCRIPTION
**Pull request by qiluo-msft**

### CODE REVIEW

File: src/ax_interface/mib.py [hunk 1]
```diff
@@ -127,10 +135,16 @@ def __init__(self, subtree, sub_ids, value_type, callable_, *args, updater=None)
 
     def __iter__(self):
         for sub_id in self.sub_ids:
-            yield self.sub_id + (sub_id,)
+            if isinstance(sub_id, tuple):
```
Comment (tombo315) [inline src/ax_interface/mib.py]: Consider moving these if/else statements out of ContextualMIBEntry into their own class (avoiding any changes to CME)

```python
class MIBSubtreeEntry(MIBEntry):
  def __init__(...):
    # impl
  def __iter__(self):
    # impl
  def __call__(self, sub_id_group)
   # impl
``` 

and remember to replace the type inspection for `ContextualMIBEntry` in the metaclass `__new__` with `MIBSubtreeEntry`

File: src/sonic_ax_impl/mibs/ietf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+def mac_decimals(mac):
```
Comment (tombo315) [inline src/sonic_ax_impl/mibs/ietf/__init__.py]: consider moving this into `ax_interface/util.py`

File: src/sonic_ax_impl/mibs/ietf/rfc4363.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            ent = self.db_conn.get_all(mibs.ASIC_DB, s, blocking=True)
+            port_oid = ent[b"SAI_FDB_ENTRY_ATTR_PORT_ID"]
+            if port_oid.startswith(b"oid:0x"):
+                port_oid = port_oid[6:]
+
+            self.vlanmac_ifindex_map[fdb_vlanmac(fdb)] = mibs.get_index(self.if_id_map[port_oid])
+
+
+    def fdb_ifindex(self, sub_id):
+        assert sub_id is not None
```
Comment (tombo315) [inline src/sonic_ax_impl/mibs/ietf/rfc4363.py]: consider using `mibs.logger` instead of `assert`.

File: tests/test_fdb.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+from ax_interface.mib import MIBTable
+from ax_interface.pdu import PDU, PDUHeader, PDUHeaderTags, supported_pdus, ContextOptionalPDU, _ignored_pdus, PDUStream
+from ax_interface.pdu_implementations import OpenPDU, ResponsePDU, RegisterPDU, GetPDU
+from ax_interface import exceptions, ValueType
+from ax_interface.encodings import ObjectIdentifier
+from ax_interface.constants import PduTypes
+from sonic_ax_impl.mibs import ieee802_1ab
+from sonic_ax_impl.mibs.ietf import rfc4363
+from sonic_ax_impl.main import SonicMIB
```
Comment (tombo315) [inline tests/test_fdb.py]: cleanup these imports

File: src/sonic_ax_impl/mibs/ietf/rfc4363.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
+import json
+from enum import unique, Enum
+
+from sonic_ax_impl import mibs
+from sonic_ax_impl.mibs.ietf import *
```
Comment (tombo315) [inline src/sonic_ax_impl/mibs/ietf/rfc4363.py]: I don't recommend `import *` in any production code. For the mac_decimals function (see my other comment) and:

```python
from ax_interface import ..., util
```

File: tests/test_fdb.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+class TestFdbMIB(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        class FdbMIB(rfc4363.FdbMIB):
```
Comment (tombo315) [inline tests/test_fdb.py]: I don't see the intent of this declaration.

File: src/sonic_ax_impl/mibs/ietf/rfc4363.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                port_oid = port_oid[6:]
+
+            self.vlanmac_ifindex_map[fdb_vlanmac(fdb)] = mibs.get_index(self.if_id_map[port_oid])
+
+
+    def fdb_ifindex(self, sub_id):
+        assert sub_id is not None
+        return self.vlanmac_ifindex_map[sub_id]
+
+class FdbMIB(metaclass=MIBMeta, prefix='.1.3.6.1.2.1.17.7.1.2.2.1'):
```
Comment (tombo315) [inline src/sonic_ax_impl/mibs/ietf/rfc4363.py]: update the readme with the RFC link

'Forwarding Database' https://tools.ietf.org/html/rfc4363

File: tests/test_fdb.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+        cls.lut = MIBTable(SonicMIB)
+
+    def test_print_oids(self):
+        for k in self.lut.keys():
+            print(k)
+        #mib_entry = self.lut[(1, 3, 6, 1, 2, 1, 17, 7, 1, 2, 2, 1, 2, 1000, 124, 254, 144, 128, 159, 92)]
+
+    def test_getpdu(self):
+        get_pdu = GetPDU(
```
Comment (tombo315) [inline tests/test_fdb.py]: changes were made to getnext

add an equivalent test for getnext from 
- (1, 3, 6, 1, 2, 1, 17, 7, 1, 2, 2, 1, 2)
- 1, 3, 6, 1, 2, 1, 17, 7, 1, 2, 2, 1, 3)

File: src/ax_interface/mib.py (hunk 1 above)
Comment (qiluo-msft) [inline src/ax_interface/mib.py]: Created a new MIBEntry subclass. Considering the dynamic OID feature, do some minor changes on existing CME class to facility the upper level implementation.

---
In reply to: [97460638](https://github.com/Azure/sonic-snmpagent/pull/3#discussion_r97460638) [](ancestors = 97460638)

File: src/sonic_ax_impl/mibs/ietf/rfc4363.py (hunk 3 above)
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc4363.py]: It is an 'assert' in concept. So no need to fix.

---
In reply to: [97461221](https://github.com/Azure/sonic-snmpagent/pull/3#discussion_r97461221) [](ancestors = 97461221)

**Comment by qiluo-msft** [comment]
@tombo315 This PR is still open to any comments.

**Pull request merged**